# Bug Report — 9router / Kiro model routing issue # Báo lỗi — lỗi route model 9router / Kiro

## 1) Summary / Tóm tắt
**EN:**
`kr/claude-sonnet-4.5` fails through 9router with `400 Improperly formed request`, even for a minimal OpenAI-compatible `chat/completions` payload. Other Kiro models on the same provider/account work normally.

**VI:**
`kr/claude-sonnet-4.5` bị lỗi qua 9router với `400 Improperly formed request`, ngay cả khi gửi payload `chat/completions` tối giản theo chuẩn OpenAI-compatible. Các model Kiro khác dùng cùng provider/account vẫn hoạt động bình thường.

---

## 2) Environment / Môi trường
**EN:**
- 9router version: `0.3.78`
- Access path: OpenAI-compatible endpoint via `POST /v1/chat/completions`
- Base URL: `http://127.0.0.1:20128/v1/chat/completions`
- API key used: `sk-9router`
- Provider: `kiro`
- Models tested:
  - `kr/claude-sonnet-4.5`
  - `kr/claude-haiku-4.5`
  - `kr/deepseek-3.2`
  - `kr/deepseek-3.1`
  - `kr/qwen3-coder-next`

**VI:**
- Phiên bản 9router: `0.3.78`
- Đường gọi: OpenAI-compatible endpoint qua `POST /v1/chat/completions`
- Base URL: `http://127.0.0.1:20128/v1/chat/completions`
- API key dùng để test: `sk-9router`
- Provider: `kiro`
- Các model đã test:
  - `kr/claude-sonnet-4.5`
  - `kr/claude-haiku-4.5`
  - `kr/deepseek-3.2`
  - `kr/deepseek-3.1`
  - `kr/qwen3-coder-next`

---

## 3) Reproduction / Cách tái hiện
**EN:**
Minimal test request:

```bash
BASE_URL="http://127.0.0.1:20128/v1/chat/completions"
API_KEY="sk-9router"

curl -sS "$BASE_URL" \
  -H "Authorization: Bearer $API_KEY" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "kr/claude-sonnet-4.5",
    "messages": [
      {"role": "user", "content": "Reply with exactly: OK"}
    ],
    "stream": false
  }'
```

**Actual result:**
```json
{"error":{"message":"[kiro/claude-sonnet-4.5] [400]: Improperly formed request. (reset after 1m 6s)"}}
```

**VI:**
Lệnh test tối giản:

```bash
BASE_URL="http://127.0.0.1:20128/v1/chat/completions"
API_KEY="sk-9router"

curl -sS "$BASE_URL" \
  -H "Authorization: Bearer $API_KEY" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "kr/claude-sonnet-4.5",
    "messages": [
      {"role": "user", "content": "Reply with exactly: OK"}
    ],
    "stream": false
  }'
```

**Kết quả thực tế:**
```json
{"error":{"message":"[kiro/claude-sonnet-4.5] [400]: Improperly formed request. (reset after 1m 6s)"}}
```

---

## 4) Control tests / Đối chứng
**EN:**
The same minimal request pattern works for the following Kiro models:

### `kr/claude-haiku-4.5`
```json
{"id":"chatcmpl-1775414267664","object":"chat.completion","created":1775414267,"model":"claude-haiku-4.5","choices":[{"index":0,"message":{"role":"assistant","content":"OK"},"finish_reason":"stop"}],"usage":{"prompt_tokens":4078,"completion_tokens":1,"total_tokens":4079}}
```

### `kr/deepseek-3.2`
```json
{"id":"chatcmpl-1775414279897","object":"chat.completion","created":1775414279,"model":"deepseek-3.2","choices":[{"index":0,"message":{"role":"assistant","content":"OK"},"finish_reason":"stop"}],"usage":{"prompt_tokens":4347,"completion_tokens":1,"total_tokens":4348}}
```

### `kr/qwen3-coder-next`
```json
{"id":"chatcmpl-1775414304317","object":"chat.completion","created":1775414304,"model":"qwen3-coder-next","choices":[{"index":0,"message":{"role":"assistant","content":"OK"},"finish_reason":"stop"}],"usage":{"prompt_tokens":3504,"completion_tokens":1,"total_tokens":3505}}
```

**VI:**
Cùng kiểu request tối giản như trên hoạt động bình thường với các model Kiro sau:

### `kr/claude-haiku-4.5`
```json
{"id":"chatcmpl-1775414267664","object":"chat.completion","created":1775414267,"model":"claude-haiku-4.5","choices":[{"index":0,"message":{"role":"assistant","content":"OK"},"finish_reason":"stop"}],"usage":{"prompt_tokens":4078,"completion_tokens":1,"total_tokens":4079}}
```

### `kr/deepseek-3.2`

```json
{"id":"chatcmpl-1775414279897","object":"chat.completion","created":1775414279,"model":"deepseek-3.2","choices":[{"index":0,"message":{"role":"assistant","content":"OK"},"finish_reason":"stop"}],"usage":{"prompt_tokens":4347,"completion_tokens":1,"total_tokens":4348}}
```

### `kr/qwen3-coder-next`
```json
{"id":"chatcmpl-1775414304317","object":"chat.completion","created":1775414304,"model":"qwen3-coder-next","choices":[{"index":0,"message":{"role":"assistant","content":"OK"},"finish_reason":"stop"}],"usage":{"prompt_tokens":3504,"completion_tokens":1,"total_tokens":3505}}
```

---

## 5) Additional finding / Phát hiện thêm
**EN:**
`kr/deepseek-3.1` fails with a different error:

```json
{"error":{"message":"[kiro/deepseek-3.1] [400]: Invalid model. Please select a different model to continue. (reset after 27s)"}}
```

This likely indicates an invalid or unsupported model ID rather than the same formatting issue.

**VI:**
`kr/deepseek-3.1` lỗi theo kiểu khác:

```json
{"error":{"message":"[kiro/deepseek-3.1] [400]: Invalid model. Please select a different model to continue. (reset after 27s)"}}
```

Lỗi này có vẻ là do model ID không hợp lệ hoặc upstream không hỗ trợ model đó, khác với lỗi format của Sonnet.

---

## 6) Relevant logs / Log liên quan
**EN:**
9router log for the failing Sonnet route:

```text
[PENDING] START | provider=kiro | model=claude-sonnet-4.5
[REQUEST] KIRO | claude-sonnet-4.5 | 0 msgs
[PENDING] END (ERROR) | provider=kiro | model=claude-sonnet-4.5
[ERROR] [400]: Improperly formed request.
❌ kiro [400]: [400]: Improperly formed request.
```

Important observation:
- the request is reaching Kiro
- but 9router logs show `0 msgs` for `claude-sonnet-4.5`
- this strongly suggests a model-specific request translation / passthrough problem

**VI:**
Log 9router khi route sang Sonnet bị lỗi:

```text
[PENDING] START | provider=kiro | model=claude-sonnet-4.5
[REQUEST] KIRO | claude-sonnet-4.5 | 0 msgs
[PENDING] END (ERROR) | provider=kiro | model=claude-sonnet-4.5
[ERROR] [400]: Improperly formed request.
❌ kiro [400]: [400]: Improperly formed request.
```

Điểm quan trọng:
- request đã đi tới được Kiro
- nhưng log 9router lại hiện `0 msgs` với `claude-sonnet-4.5`
- điều này gợi ý rất mạnh rằng có lỗi ở tầng request translation / passthrough riêng của model Sonnet

---

## 7) Expected behavior / Hành vi mong đợi
**EN:**
`kr/claude-sonnet-4.5` should accept the same minimal OpenAI-compatible `chat/completions` payload pattern that works for other Kiro models.

**VI:**
`kr/claude-sonnet-4.5` nên chấp nhận cùng kiểu payload tối giản theo chuẩn OpenAI-compatible `chat/completions`, giống như các model Kiro khác đang chạy được.

---

## 8) Hypothesis / Giả thuyết lỗi
**EN:**
Likely one of:
1. model-specific request translator issue for `claude-sonnet-4.5`
2. passthrough layer dropping messages for this route
3. compatibility mismatch specific to Kiro Sonnet handling

This does **not** look like:
- WAN IP blocking
- global Kiro outage
- invalid API key
- bad prompt content

**VI:**
Khả năng cao là một trong các trường hợp:
1. lỗi request translator riêng cho `claude-sonnet-4.5`
2. lớp passthrough làm rơi `messages` ở route này
3. mismatch compatibility riêng cho cách 9router xử lý Kiro Sonnet

Lỗi này **không giống**:
- bị chặn IP WAN
- Kiro chết toàn cục
- API key sai
- prompt có nội dung xấu

## 9) Request / Đề nghị kiểm tra
**EN:**
Please verify the Kiro route/adapter for `claude-sonnet-4.5`, especially:
- message passthrough / translator behavior
- whether `messages[]` are dropped before upstream send
- model-specific compatibility handling compared with `claude-haiku-4.5`

**VI:**
Nhờ kiểm tra route/adapter Kiro cho `claude-sonnet-4.5`, đặc biệt là:
- hành vi passthrough / translator của messages
- có bị drop `messages[]` trước khi gửi upstream không
- cách xử lý compatibility riêng của model này so với `claude-haiku-4.5`

--- a/open-sse/translator/request/openai-to-kiro.js
+++ b/open-sse/translator/request/openai-to-kiro.js
@@ -254,6 +254,20 @@ function convertMessages(messages, tools, model) {
     }
   });
 
+  // Merge consecutive user messages (Kiro requires alternating user/assistant)
+  const mergedHistory = [];
+  for (let i = 0; i < history.length; i++) {
+    const current = history[i];
+    if (current.userInputMessage &&
+        mergedHistory.length > 0 &&
+        mergedHistory[mergedHistory.length - 1].userInputMessage) {
+      const prev = mergedHistory[mergedHistory.length - 1];
+      prev.userInputMessage.content += "\n\n" + current.userInputMessage.content;
+    } else {
+      mergedHistory.push(current);
+    }
+  }
+
   // Inject tools into currentMessage AFTER cleanup
   if (firstHistoryTools && currentMessage?.userInputMessage &&
       !currentMessage.userInputMessage.userInputMessageContext?.tools) {
@@ -263,7 +277,7 @@ function convertMessages(messages, tools, model) {
     currentMessage.userInputMessage.userInputMessageContext.tools = firstHistoryTools;
   }
 
-  return { history, currentMessage };
+  return { history: mergedHistory, currentMessage };
 }
 
 /**

